### PR TITLE
Implement MetricDetails

### PR DESCRIPTION
**Pull request by jessie-ross**
## ~~This is a WIP PR welcoming suggestions.~~

**A minimalist implementation of MetricDetails:**

Screen Shot 2020-06-29 at 1 13 16 am

- Uses the MaterialTable DetailsView feature
- Prints the parameters as JSON, I think this is reasonable for the MVP.

### Known issues to be fixed in PR:
- Testing
- Styling 
  - I know it doesn't look pretty now, but it can be polished up real nice. 
  - Alignment, background and scroll-overflow.

## How has this been tested?
- AT & MT

### CODE REVIEW

File: components/MetricsTable.tsx [hunk 1]
```diff
@@ -14,22 +15,73 @@ const tableColumns = [
...
+  return (
+    <>
+      {isLoading && <LinearProgress />}
+      <Snackbar open={!!error} message='Oops! Something went wrong while trying to load a Metric.' />
+      {!isLoading && !error && metricFull && (
+        <dl className={classes.root}>
+          <Typography component='dt'>Direction:</Typography>
+          <Typography component='dd'>{metricFull.higherIsBetter ? 'Higher is better.' : 'Lower is better.'}</Typography>
+          <Typography component='dt'>Parameters:</Typography>
+          <Typography component='dd'>{paramsStringified}</Typography>
```
Comment (yanirs) [inline components/MetricsTable.tsx:63]: I think it'd be better to render the parameters as more user-friendly text or not render them at all. As noted in #162, the idea is to make it a standalone page, so we have more space to play with, though I'm not too opposed to using the detail panel.

File: components/MetricsTable.tsx [hunk 2]
```diff
@@ -14,22 +15,73 @@ const tableColumns = [
...
+  const paramsStringified = params && JSON.stringify(params, null, 4)
+
+  return (
+    <>
+      {isLoading && <LinearProgress />}
+      <Snackbar open={!!error} message='Oops! Something went wrong while trying to load a Metric.' />
+      {!isLoading && !error && metricFull && (
+        <dl className={classes.root}>
+          <Typography component='dt'>Direction:</Typography>
+          <Typography component='dd'>{metricFull.higherIsBetter ? 'Higher is better.' : 'Lower is better.'}</Typography>
```
Comment (yanirs) [inline components/MetricsTable.tsx:61]: Can make this `Higher is better?` with `Yes/No` in the `dd`, and then we can use the `formatBoolean()` util.

File: components/MetricsTable.tsx [hunk 3]
```diff
@@ -14,22 +15,73 @@ const tableColumns = [
...
 
+const useMetricDetailStyles = makeStyles((theme: Theme) =>
+  createStyles({
+    root: {
+      display: 'grid',
+      gridTemplateColumns: '10rem auto',
+      gridRowGap: theme.spacing(1),
+      padding: theme.spacing(1, 4),
+      '& dt': {
+        fontWeight: 600,
```
Comment (yanirs) [inline components/MetricsTable.tsx:26]: As noted on the other PR, this 600 should probably be a constant somewhere. :slightly_smiling_face:

File: components/MetricsTable.tsx (hunk 1 above)
Comment (jessie-ross) [reply components/MetricsTable.tsx:63]: Yeah I think it is worth doing it like this until other higher priority items are ticked off.
(If this were to load into its own page it would look very empty at the moment.)

**Comment by jessie-ross** [comment]
I am holding on this one until #170 goes through, as I can apply a lot of that here.

**Comment by jessie-ross** [comment]
The latest look https://d.pr/v/Ca1xfS (I do like including these...)

- In a sub-grid with hover-colour background.
- Contained the JSON visually with a border and an `overflow: scroll`.

File: components/MetricsTable.tsx [hunk 4]
```diff
@@ -1,35 +1,141 @@
...
 import { defaultTableOptions } from '@/utils/material-table'
 
 const debug = debugFactory('abacus:components/MetricsTable.tsx')
 
-const tableColumns = [
-  { title: 'Name', field: 'name' },
-  { title: 'Description', field: 'description' },
-  { title: 'Parameter Type', field: 'parameterType' },
-]
+/* istanbul ignore next */
```
Comment (yanirs) [inline components/MetricsTable.tsx:25]: Having so many coverage suppression comments defeats the purpose of checking for coverage. I'm happy to reopen the testing discussion on P2, but if the entire file is ignored, let's suppress on the config level rather than sprinkle suppression comments everywhere.

File: templates/component.tsx [hunk 5]
```diff
@@ -7,6 +7,7 @@
 import React from 'react'
 import { makeStyles, createStyles, Theme } from '@material-ui/core/styles'
 
+/* istanbul ignore next */
```
Comment (yanirs) [inline templates/component.tsx:10]: Is this meant to be in this PR?

**Comment by yanirs** [comment]
> The latest look https://d.pr/v/Ca1xfS (I do like including these...)

Nice video! Looks good. I suppose it's no longer WIP?

I'm not a fan of suppressing all coverage, but I see how this would be tricky to test well now that we rely more heavily on MUI components and MaterialTable magic. I'm happy to suppress for now, but as noted, let's do it once per file rather than sprinkle warnings everywhere, and aim to have the discussion around testing and coverage moving forward early next week.

File: templates/component.tsx (hunk 5 above)
Comment (jessie-ross) [reply templates/component.tsx:10]: Yep, code coverage is complaining about the function being created here, I couldn't find a way to turn it off

File: components/MetricsTable.tsx (hunk 4 above)
Comment (jessie-ross) [reply components/MetricsTable.tsx:25]: Yep, code coverage is complaining about the style function. I think they shouldn't be tested.

It is just this function and not the whole file, e.g. with the current decision around testing, if another function were made we should be testing that.

Comment (yanirs) [reply components/MetricsTable.tsx:25]: There are three `istanbul ignore next` comments in this file, which seem to pretty much cover everything.

**Comment by jessie-ross** [comment]
> Nice video! Looks good. I suppose it's no longer WIP?

Good point, I have just updated the title.

> I'm not a fan of suppressing all coverage

Firstly not all coverage is being suppressed at the moment, e.g. this component:
https://github.com/Automattic/abacus/blob/f0f11ce2ce5caa2c4ac9abe5a638877b9c24790b/components/MetricsTable.tsx#L97-L100

Secondly, the parts we are ignoring fall into three specific categories:
- Trivial and unnecessary - ignoring the theme injecting function in makeStyles.
   I couldn't figure out a way to permanently ignore all of these.
- Incredibly trivial functions that are covered by e2e, e.g. these two arrow functions:
https://github.com/Automattic/abacus/blob/f0f11ce2ce5caa2c4ac9abe5a638877b9c24790b/components/MetricsTable.tsx#L129-L138
- Side-effectful data loading component that is covered by e2e, e.g. this fella:
https://github.com/Automattic/abacus/blob/f0f11ce2ce5caa2c4ac9abe5a638877b9c24790b/components/MetricsTable.tsx#L53-L56

I am turning off coverage for these specific purposes but I don't necessarily want to turn it off in general, as it fits to have it on with this codebase.

File: templates/component.tsx (hunk 5 above)
Comment (dhurlburtusa) [reply templates/component.tsx:10]: This is bizarre. We are already explicitly excluding all matching files in the `templates/` directory. See `jest.config.js`, line 14. We shouldn't need this.

Comment (yanirs) [reply templates/component.tsx:10]: I tried removing it locally and I'm not seeing any issues.

File: components/MetricsTable.tsx (hunk 4 above)
Comment (yanirs) [reply components/MetricsTable.tsx:25]: I played with it a bit locally, and simply adding a click on a row to the unit tests obviates the need for all these coverage suppression comments, and moves the attention to things that aren't tested but should be.

**Comment by yanirs** [comment]
> Firstly not all coverage is being suppressed at the moment, e.g. this component:

Ah yeah, I see that now.

> Secondly, the parts we are ignoring fall into three specific categories:
>
> Trivial and unnecessary - ignoring the theme injecting function in makeStyles.
> I couldn't figure out a way to permanently ignore all of these.

We shouldn't need to permanently ignore all these. It's easy to cover them by rendering the component. As noted inline, in this case adding a row click to the existing test would render the `MetricDetail` component. Alternatively, we can pull it out and render it separately.

> Incredibly trivial functions that are covered by e2e, e.g. these two arrow functions:

These functions are very trivial to cover by adding a row click to the test. :slightly_smiling_face: I think it's worth doing in this case.

> Side-effectful data loading component that is covered by e2e, e.g. this fella:

There's plenty of behaviour in `MetricDetail` that I think is worth testing. It shouldn't be too hard to do if we mock the data loading line.

**Comment by jessie-ross** [comment]
Ah ok I see what you mean, I am not too familiar with Unit testing React Components, I will add the click to the unit test.

**Comment by jessie-ross** [comment]
> There's plenty of behaviour in MetricDetail that I think is worth testing. It shouldn't be too hard to do if we mock the data loading line.

@yanirs, it seems like all the rest of the data loading is currently being done in `/pages` with only e2e applied so I can't seem to see any examples of mocking api hooks, I am going to look into a good way to do this.

**Comment by jessie-ross** [comment]
This is actually pulling on a long thread:
- Mocking the Api call wasn't bad
- Getting the fixture data is complicated/lengthy


I didn't have a MetricFull fixture so I went to create one. I wanted to extend from MetricBare, so I thought about using the MetricFull constructor, but I couldn't use it directly as I would have to strip the methods first. The I thought about using  MetricBare.toApiData and MetricFull.fromApiData together, and I found out toApiData doesn't exist on MetricBare. I also wanted to keep it consistent with the rest of the code in the file which was just using Model constructors. At this point I was like this isn't worth it and copy-pasted it in.

Then I realised to mock the Api call I needed api data, so I was thinking I would just use MetricFull.toApiData, which didn't exist. At this point I thought it would be a good idea to write it so I did, but then I realised I would have to do the same for the children entities, so I did that.

Ok, so I thought I was all good at this point but nope, the code coverage doesn't pass now. I didn't want to do half assed testing if I was going to be testing these but these are lengthy to test well because of all the variations so I decided I should stop here.

I would say #204 would fix a lot of these problems.
- Integration testing assuring the schemas
- Only need one set of fixtures - the server-side data

Then it is easy to extend/convert the fixtures, and they are all flowing in one direction from the source data.

**Comment by jessie-ross** [comment]
I've realised this morning that the mocked data doesn't need to be in ApiData format 🤦

File: components/MetricsTable.tsx [hunk 6]
```diff
@@ -1,35 +1,141 @@
...
-    }
-    router.push('/metrics/[id]', `/metrics/${rowData.metricId}`)
-  }
+  const theme = useTheme()
+  const tableColumns = [
+    {
+      title: 'Name',
+      field: 'name',
+      cellStyle: {
+        fontFamily: theme.custom.fonts.monospace,
```
Comment (yanirs) [inline components/MetricsTable.tsx:110]: Doesn't matter for this PR, but can't we just apply the monospace font higher up in the CSS? This is getting a bit repetitive.

File: models/Event.ts [hunk 7]
```diff
@@ -12,7 +12,7 @@ export class Event {
   /**
    * Event properties that further constrain this event selection.
    */
-  public readonly props?: { [key: string]: string }
+  public readonly props?: { [key: string]: string | boolean }
```
Comment (yanirs) [inline models/Event.ts:15]: This should be a mapping from string to string since that's how `eventprops` are stored by Tracks.

File: models/MetricRevenueParams.ts [hunk 8]
```diff
@@ -4,6 +4,11 @@ import { ApiData } from '@/api/ApiData'
  * Parameters for a revenue query. If `null`, then `event_params` must be given.
  */
 export class MetricRevenueParams {
+  /**
+   * Description
+   */
+  public readonly description?: string
```
Comment (yanirs) [inline models/MetricRevenueParams.ts:10]: Why was this added? We have descriptions on the metric level, not on the params level.

Comment (jessie-ross) [reply models/MetricRevenueParams.ts:10]: This was [based off of the swagger docs](https://app.swaggerhub.com/apis/yanir/experiments/0.1.0#/MetricFull), if you expand the revenueParams section.

File: models/Event.ts (hunk 7 above)
Comment (jessie-ross) [reply models/Event.ts:15]: [Also based off of swagger](https://app.swaggerhub.com/apis/yanir/experiments/0.1.0#/management/get_metrics__metric_id_), if you look at the 200 example.

**Comment by jessie-ross** [comment]
I'll make the schema changes you mentioned and add an issue for now about updating the swagger docs.

I will likely make the change to the swagger docs along with changes needed for using Schema in the codebase.

File: models/MetricRevenueParams.ts (hunk 8 above)
Comment (yanirs) [reply models/MetricRevenueParams.ts:10]: Looks like it's just the way Swagger UI formats things:

image

`description:` is slightly greyed out, indicating that it's _not_ a property. It's clearer when looking at the yaml:

```yaml
revenue_params:
  type: object
  description: >
    Parameters for a revenue query. If `null`, then
    `event_params` must be given.
  nullable: true
  properties:
    refund_days:
      type: integer
      description: Number of days to use for the refund window.
      format: int32
...
```

They should really render it like they render the array description. It's pretty confusing.

File: models/Event.ts (hunk 7 above)
Comment (yanirs) [reply models/Event.ts:15]: Ah yeah, that's a bug in the conversion metric example, so it should be fixed in the Swagger file.

File: components/MetricsTable.tsx (hunk 6 above)
Comment (jessie-ross) [reply components/MetricsTable.tsx:110]: I've tried to do so but it seems to be a limitation of MaterialGrid, part of my motivation for #205

File: models/MetricRevenueParams.ts (hunk 8 above)
Comment (jessie-ross) [reply models/MetricRevenueParams.ts:10]: Ahhh I see
